Add functions to generate QC metrics for benchmarking

allyhawkins opened the pull request:
In working on #92 and the `scpcaTools` package, I also started working on ways to grab the output data from s3, import it using `scpcaTools`, calculate QC metrics, and create the dataframes needed for plotting the comparisons we are interested in for benchmarking. We have done this a few different ways now and previously had been using the code in `01-import-quant-data.Rmd`. In that script we import data for each unique tool and unique parameter combination to create a separate list of `SingleCellExperiments` before combining the `colData` or `rowData` altogether into one dataframe used for generating plots of interest. 

As we have started to repeat some tasks present in that code a few times now, it made sense to me to start splitting out some of the more specific benchmarking and internal data wrangling related steps into functions. So here I am adding in a group of benchmarking functions: 

1. `aws_copy_samples.R`: This function takes a list of samples, a location on S3, and the tools used to grab all the relevant samples from S3 and copy over to a local directory. 
2. `quant_info_table.R`: This creates a table with relevant information about each sample that was run and with what tool configuration that was used. It also creates the necessary columns needed to run `scpcaTools::import_quant_data()` based on what tool and parameters were used. 
3. `make_sce_list.R`: This function applies `scpcaTools::import_quant_data()` to a set of samples and creates a list of `SingleCellExperiments` and then adds both `colData` and `rowData` using `scater::addPerCellQC` and `scater::addPerFeatureQC`. 
4. `coldata_to_df` and `rowdata_to_df` both will take the row or colData of a `SingleCellExperiment` object and create a dataframe. 

Using these 4 functions (some may or may not be useful things we could think about adding to scpcaTools), I then have a main script, `benchmarking_generate_qc_df.R` that takes a list of samples and tools used, creates lists of `SingleCellExperiment` objects and saves them as `.rds` files, and then writes out the `quant_info`, coldata dataframe and rowdata dataframe for all samples used in the benchmarking experiment. These output files are similar to the tables that were the inputs for all of the plots that were generated in `04-cell-level-benchmarking-metrics.Rmd` and `05-gene-level-benchmarking-metrics.Rmd`. My thought process behind making this a script is to be able to use this to create these same dataframes and `.rds` files for any groupings of samples and tools that we might be interested in looking at. 

In addition to getting feedback on overall flow and structure of the main script interacting with the additional functions, I have a few more specific questions for reviewers: 
- For alevin-fry, I chose to break up the importing by configuration of alevin-fry, something you don't need for other tools - do we think this is necessary or could we treat it all as one and save one large `.rds` file? 
- Are there any places where things don't belong here but belong in `scpcaTools` instead?
- are any functions extraneous and need to be removed entirely?

jashapiro reviewed (commented):
Okay, so there is a lot here! In the future, I might prefer if this were broken up into a few separate PRs to facilitate focused discussion. Maybe separate the download functions into one PR and the processing into a separate one.

Also on that front: comments! Especially when you get to making lists of lists and such, it will really help to know what is data is expected at each stage. Without that, it was sometimes hard to figure out what you were trying to do at various points. Particularly for functions, having a set of comments describing input and output is super helpful.

So that said, I worked my way through what I felt I could. There are some obvious candidates for `scpcaTools` that I noted: pulling out data frames is a thing we seem to do frequently, so we should just add those to the package.

My other comments are mostly below, but a couple general notes:

I tried to convert some of your `for` loops to `purrr` calls. Basically any time I saw a counter `i` I wanted to get rid of it, mostly because those are a common source of errors. I didn't make every change though, mostly to limit the number of comments somewhat... I think I caught the general use cases though.

allyhawkins commented:
Thanks for taking the time to go through this @jashapiro, I really appreciate it and apologize for making it into such a large PR with so many moving pieces. I think I was still working through how everything would fit together and was having trouble explaining one piece without the other, probably because of my lack of comments! 

I'll go through and add in your comments, move some functions over to `scpcaTools`, add in a lot more comments, and clean up the `for` loops. I need to get more comfortable using `purrr` and all of its capabilities!

jashapiro commented on analysis/benchmarking-functions/R/quant_info_table.R:
```diff
@@ -53,10 +53,7 @@ quant_info_table <- function(data_dir, tools, samples){
-                  intron_mode = ifelse(tool != "cellranger" & 
-                                         usa_mode == FALSE & 
-                                         index_type == "splici", TRUE, FALSE))
+                  usa_mode = index_type == "splici" & alevin_resolution == "cr" &  
```
I had an extra `&` in there... would get caught quickly, but I noticed it now.
```suggestion
                  usa_mode = index_type == "splici" & alevin_resolution == "cr" 
```

jashapiro commented on analysis/benchmarking-functions/R/make_sce_list.R:
```diff
@@ -0,0 +1,24 @@
+addCellQC <- function(sce, mito, threshold = 0){
```
One other comment on this function, which came up when I was looking at https://github.com/AlexsLemonade/alsf-scpca/pull/101/files#r668161413 and I would think should be addressed especially if we add it to `scpcaTools`

The function name here follows the convention of `scater`, but I think that is potentially confusing in this case, because it does not match the other function names. We should follow the same naming convention for all of our functions. 

I also would prefer if the function name gave more of a hint of how it differs from `scater::addPerCellQC`. I would suggest a name like `add_mito_cell_qc` to indicate that it specifies the mitochondrial genes as a subset.

allyhawkins commented:
I have gone through and taken out a the functions that calculate per cell qc with a mito subset and the functions that grab `colData` and `rowData` and put them into dataframes and transferred those over to scpcaTools. The remaining functions, I have simplified, added in documentation using the `roxygen2` skeleton to keep documentation consistent, and added in more comments. For the main script, I also have tried to simplify it, removing all of the `for` loops with counters. I made saving the `.rds` files an optional argument using the `--save` flag in the main R script. For input to the script, it now takes two `.tsv` files, one is a list of sample IDs and one is a list of tools used. I have also tried to incorporate some more checks along the way in both the main script and the functions. 

In reviewing, please let me know if there are still places that are unclear, and/or places that could still be further simplified or done in a more efficient manner.

allyhawkins commented on analysis/benchmarking-functions/R/aws_copy_samples.R:
```diff
@@ -0,0 +1,38 @@
+    dir.create(local_tool_dir, recursive = TRUE, showWarnings = FALSE)
+    
+    if (tool == 'alevin') {
+      sync_call <- paste('aws s3 cp', s3_tool_dir, local_tool_dir, 
```
So there is a reason I chose to switch it over to using `cp` instead and that's because last time when I was using `sync`, it wasn't recognizing that when I had an old version of a sample that was not run fully it didn't copy over it (that was the issue with that weird sample last time), even though the files were different. So to make sure that we are always working with the correct files from S3 from that sample and don't have that mix up, switching to `cp` made the most sense to me.

jashapiro replied:
Ah yes, that was annoying. Seems a reasonable change.

jashapiro reviewed (approved):
This looks good to me. The comments make it much easier to follow along. I had one place to DRY up a bit, but other than that it looks like it should be fine for finishing up the benchmarking.

allyhawkins commented on analysis/benchmarking-functions/R/make_sce_list.R:
```diff
@@ -0,0 +1,39 @@
+#'
+make_sce_list <- function(info_df, mito, filename = NULL){
+  # for each row in the info data frame, create a SingleCellExperiment object and save as a list
+  sce_list <- mapply(scpcaTools::import_quant_data,
```
yea I had played around with using that and some other options and I didn't like it very much so had chosen to take this route instead.
